programs/dynamic-bonding-curve/src/state/virtual_pool.rs

```diff
@@ -204,31 +205,135 @@ impl VirtualPool {
     }
 
-    pub fn get_swap_result(
+    pub fn get_swap_exact_out_result(
         &self,
         config: &PoolConfig,
-        amount_in: u64,
+        amount_out: u64,
         fee_mode: &FeeMode,
         trade_direction: TradeDirection,
         current_point: u64,
+        rate_limiter: Option<&FeeRateLimiter>,
     ) -> Result<SwapResult> {
         let mut actual_protocol_fee = 0;
         let mut actual_trading_fee = 0;
         let mut actual_referral_fee = 0;
 
-        let actual_amount_in = if fee_mode.fees_on_input {
+        let trade_fee_numerator = config.pool_fees.get_total_trading_fee(
+            &self.volatility_tracker,
+            current_point,
+            self.activation_point,
+            0, // Rate limiter applied later on quote to base and collect fee quote mode
+            trade_direction,
+        )?;
+
+        let included_fee_out_amount = if fee_mode.fees_on_input {
+            amount_out
+        } else {
+            let included_fee_out_amount =
+                PoolFeesConfig::get_included_fee_amount(trade_fee_numerator, amount_out)?;
+            let FeeOnAmountResult {
+                protocol_fee,
+                trading_fee,
+                referral_fee,
+                ..
+            } = config.pool_fees.get_fee_on_amount(
+                trade_fee_numerator,
+                included_fee_out_amount,
+                fee_mode.has_referral,
+            )?;
+            actual_protocol_fee = protocol_fee;
+            actual_trading_fee = trading_fee;
+            actual_referral_fee = referral_fee;
+            included_fee_out_amount
+        };
+
+        let SwapAmount {
+            amount_in,
+            next_sqrt_price,
+            ..
+        } = match trade_direction {
+            TradeDirection::BaseToQuote => {
+                self.get_swap_out_amount_from_base_to_quote(config, included_fee_out_amount)?
+            }
+            TradeDirection::QuoteToBase => {
+                self.get_swap_out_amount_from_quote_to_base(config, included_fee_out_amount)?
+            }
+        };
+
+        let in_amount = if fee_mode.fees_on_input {
+            let included_fee_in_amount = match rate_limiter {
+                Some(rate_limiter)
+                    if rate_limiter.is_rate_limiter_applied(
+                        current_point,
+                        self.activation_point,
+                        trade_direction,
+                    )? =>
+                {
+                    rate_limiter.get_included_fee_amount(amount_in)?
+                }
+                _ => PoolFeesConfig::get_included_fee_amount(trade_fee_numerator, amount_in)?,
+            };
+
             let FeeOnAmountResult {
-                amount,
                 protocol_fee,
                 trading_fee,
                 referral_fee,
+                amount,
             } = config.pool_fees.get_fee_on_amount(
-                &self.volatility_tracker,
+                trade_fee_numerator,
```

Comment:
use trade fee numerator here is incorrect, because if rate limiter is applied, the actual `trade_fee_numerator ` would be different

Reply:
Fixed. Good catch.